src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.CoreRT.cs

```diff
@@ -0,0 +1,476 @@
+    {
+        internal static IntPtr DefaultIUnknownVftblPtr { get; } = CreateDefaultIUnknownVftbl();
+
+        internal static Guid IID_IUnknown = new Guid(0x00000000, 0x0000, 0x0000, 0xc0, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x46);
+
+        private static readonly ConditionalWeakTable<object, ManagedObjectWrapperHolder> CCWTable = new ConditionalWeakTable<object, ManagedObjectWrapperHolder>();
+
+        /// <summary>
+        /// ABI for function dispatch of a COM interface.
+        /// </summary>
+        public unsafe partial struct ComInterfaceDispatch
+        {
+            /// <summary>
+            /// Given a <see cref="System.IntPtr"/> from a generated Vtable, convert to the target type.
+            /// </summary>
+            /// <typeparam name="T">Desired type.</typeparam>
+            /// <param name="dispatchPtr">Pointer supplied to Vtable function entry.</param>
+            /// <returns>Instance of type associated with dispatched function call.</returns>
+            public static unsafe T GetInstance<T>(ComInterfaceDispatch* dispatchPtr) where T : class
+            {
+                ManagedObjectWrapper* comInstance = ToManagedObjectWrapper(dispatchPtr);
+                return Unsafe.As<T>(RuntimeImports.RhHandleGet(comInstance->Target));
+            }
+
+            internal static unsafe ManagedObjectWrapper* ToManagedObjectWrapper(ComInterfaceDispatch* dispatchPtr)
+            {
+                return ((InternalComInterfaceDispatch*)dispatchPtr)->thisPtr;
+            }
+        }
+
+        internal unsafe struct InternalComInterfaceDispatch
+        {
+            public IntPtr Vtable;
+            internal ManagedObjectWrapper* thisPtr;
+        }
+
+        internal enum CreateComInterfaceFlagsEx
+        {
+            None = CreateComInterfaceFlags.None,
+            CallerDefinedIUnknown = CreateComInterfaceFlags.CallerDefinedIUnknown,
+            TrackerSupport = CreateComInterfaceFlags.TrackerSupport,
+
+            // Highest bits are reserved for internal usage
+            LacksICustomQueryInterface = 1 << 29,
+            IsComActivated = 1 << 30,
+            IsPegged = 1 << 31,
+
+            InternalMask = IsPegged | IsComActivated | LacksICustomQueryInterface,
+        }
+
+        internal unsafe struct ManagedObjectWrapper
+        {
+            public volatile IntPtr Target; // This is GC Handle
+            public long RefCount;
+
+            public int UserDefinedCount;
+            public ComInterfaceEntry* UserDefined;
+            internal InternalComInterfaceDispatch* Dispatches;
+
+            internal volatile CreateComInterfaceFlagsEx Flags;
+            const ulong ComRefCountMask = 0x000000007fffffffUL;
+            static uint GetComCount(ulong c)
+            {
+                return (uint)(c & ComRefCountMask);
+            }
+
+            public uint AddRef()
+            {
+                return GetComCount((ulong)Interlocked.Increment(ref RefCount));
+            }
+
```

Comment:
```suggestion
            public int RefCount;

            public int UserDefinedCount;
            public ComInterfaceEntry* UserDefined;
            internal InternalComInterfaceDispatch* Dispatches;

            internal volatile CreateComInterfaceFlagsEx Flags;

            public uint AddRef()
            {
                return (uint)Interlocked.Increment(ref RefCount));
            }

```
The mask on the refcount is not used for anything in this PR. It would be best to omit it until it is actually needed.

Reply:
@jkotas do I need `volatile` on the flags? I also do not rely on that in any way.